File: server/scripts/process_images.py
import os
import io
import time
import requests
import psycopg2
import subprocess
from psycopg2.extras import DictCursor
from dotenv import load_dotenv

def remove(input_data_bytes):
    try:
        # Use local file to avoid temp path permissions/formatting issues
        unique_id = str(int(time.time() * 1000))
        tmp_in_path = os.path.abspath(f"temp_in_{unique_id}.png")
        tmp_out_path = os.path.abspath(f"temp_out_{unique_id}.png")

        with open(tmp_in_path, "wb") as f:
            f.write(input_data_bytes)

        # Call Node script
        script_path = os.path.join(os.path.dirname(__file__), "remove_bg.js")
        
        result = subprocess.run(
            ["node", script_path, tmp_in_path, tmp_out_path], 
            capture_output=True, 
            text=True
        )

        if result.returncode != 0 or "SUCCESS" not in result.stdout:
            # print(f"⚠️  Node BG removal failed: {result.stderr}") # Reduce noise
            # Log error log?
            pass
            return None

        # Read output
        if os.path.exists(tmp_out_path):
            with open(tmp_out_path, "rb") as f:
                output_data = f.read()
        else:
            return None

        # Cleanup
        if os.path.exists(tmp_in_path): os.remove(tmp_in_path)
        if os.path.exists(tmp_out_path): os.remove(tmp_out_path)

        return output_data

    except Exception as e:
        print(f"❌ Error in Node bridge: {e}")
        return None

from supabase import create_client, Client
from tqdm import tqdm
from PIL import Image

# 1. Load Environment Variables
# ------------------------------------------------------------------------------
# Try to load from ../.env first
env_path = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), '.env')
print(f"Loading env from: {env_path}")
load_dotenv(env_path)

# Fallback/Override from local .env
load_dotenv()

SUPABASE_URL = os.getenv("SUPABASE_URL")
# Try multiple common names for the key
SUPABASE_KEY = os.getenv("SUPABASE_SERVICE_ROLE_KEY") or os.getenv("SUPABASE_KEY") or os.getenv("STORAGE_KEY") or os.getenv("SUPABASE_ANON_KEY")
DATABASE_URL = os.getenv("DATABASE_URL")

print(f"DEBUG: Found SUPABASE_URL: {bool(SUPABASE_URL)}")
print(f"DEBUG: Found SUPABASE_KEY: {bool(SUPABASE_KEY)}")
print(f"DEBUG: Found DATABASE_URL: {bool(DATABASE_URL)}")

if not SUPABASE_URL or not SUPABASE_KEY or not DATABASE_URL:
    print("❌ Error: Missing environment variables. Please check .env file.")
    print("Required: SUPABASE_URL, SUPABASE_KEY, DATABASE_URL")
    exit(1)

# 2. Setup Clients
# ------------------------------------------------------------------------------
try:
    supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)
except Exception as e:
    print(f"❌ Failed to initialize Supabase client: {e}")
    exit(1)

def get_db_connection():
    try:
        # Clean DATABASE_URL for psycopg2 (remove params like pgbouncer=true)
        clean_url = DATABASE_URL.split('?')[0]
        # print(f"DEBUG: Connecting to {clean_url}")
        conn = psycopg2.connect(clean_url)
        return conn
    except Exception as e:
        print(f"❌ Failed to connect to database: {e}")
        exit(1)

# 3. Processing Functions
# ------------------------------------------------------------------------------

def download_image(url):
    """Downloads image from URL into a byte stream."""
    try:
        headers = {
            'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
        }
        response = requests.get(url, headers=headers, stream=True, timeout=10)
        response.raise_for_status()
        return io.BytesIO(response.content)
    except Exception as e:
        # print(f"⚠️ Error downloading {url}: {e}")
        return None

def process_image_rembg(image_bytes):
    """Removes background from image bytes using rembg (Node Bridge)."""
    try:
        # Convert BytesIO to bytes
        input_bytes = image_bytes.getvalue()
        output_bytes = remove(input_bytes)
        return output_bytes
    except Exception as e:
        print(f"⚠️ Error processing image with rembg: {e}")
        return None

def upload_to_supabase(bucket_name, file_name, file_data):
    """Uploads file to Supabase Storage and returns public URL."""
    try:
        # Check if file exists (optional, or just overwrite)
        # supabase.storage.from_(bucket_name).remove([file_name]) # Uncomment to force clean

        # Upload
        # file_data can be bytes
        res = supabase.storage.from_(bucket_name).upload(
            file=file_data,
            path=file_name,
            file_options={"content-type": "image/png", "upsert": "true"}
        )
        
        # Get Public URL
        public_url_res = supabase.storage.from_(bucket_name).get_public_url(file_name)
        return public_url_res
    except Exception as e:
        print(f"⚠️ Error uploading to Supabase: {e}")
        return None

# 4. Main Workflow
# ------------------------------------------------------------------------------
def main():
    print("🚀 Starting Batch Image Processing...")
    
    # Connect to DB
    conn = get_db_connection()
    cur = conn.cursor(cursor_factory=DictCursor)

    # Fetch Products
    print("🔍 Fetching products from database...")
    try:
        # Note: Using double quotes for "Product" table as Prisma typically uses case-sensitive names
        cur.execute('SELECT product_id, main_image_url FROM "Product" WHERE main_image_url IS NOT NULL')
        products = cur.fetchall()
        print(f"📦 Found {len(products)} products.")
    except Exception as e:
        print(f"❌ Error fetching products: {e}")
        conn.close()
        exit(1)

    # Process Loop
    success_count = 0
    fail_count = 0
    skip_count = 0

    BUCKET_NAME = "product"

    for product in tqdm(products, desc="Processing Images"):
        pid = product['product_id']
        original_url = product['main_image_url']

        # Skip if already processed (simple heuristic: implies it's already on our supabase)
        if SUPABASE_URL in original_url:
            skip_count += 1
            continue
        
        # 1. Download
        image_stream = download_image(original_url)
        if not image_stream:
            fail_count += 1
            continue

        # 2. Remove Background
        processed_bytes = process_image_rembg(image_stream)
        if not processed_bytes:
            fail_count += 1
            continue

        # 3. Upload to Supabase
        filename = f"processed_{pid}.png"
        new_url = upload_to_supabase(BUCKET_NAME, filename, processed_bytes)

        if new_url:
            # 4. Update Database
            try:
                update_sql = 'UPDATE "Product" SET main_image_url = %s WHERE product_id = %s'
                cur.execute(update_sql, (new_url, pid))
                conn.commit()
                success_count += 1
            except Exception as e:
                print(f"❌ Error updating DB for product {pid}: {e}")
                conn.rollback()
                fail_count += 1
        else:
            fail_count += 1
            
        # Optional: Sleep briefly to avoid rate limits if necessary
        # time.sleep(0.1)

    print("\n==========================================")
    print("✅ Processing Complete")
    print(f"🟢 Successful: {success_count}")
    print(f"🔴 Failed: {fail_count}")
    print(f"⚪ Skipped (Already processed): {skip_count}")
    print("==========================================")

    cur.close()
    conn.close()

if __name__ == "__main__":
    main()

File: server/scripts/process_images.js
import { removeBackground } from "@imgly/background-removal-node";
import { createClient } from '@supabase/supabase-js';
import { PrismaClient } from '@prisma/client';
import dotenv from 'dotenv';
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import cliProgress from 'cli-progress';

// Setup paths
const __dirname = path.dirname(fileURLToPath(import.meta.url));

// Load Env
const parentEnvPath = path.resolve(__dirname, '../../.env');
if (fs.existsSync(parentEnvPath)) {
    dotenv.config({ path: parentEnvPath });
}
dotenv.config();

// Config
const SUPABASE_URL = process.env.SUPABASE_URL;
const SUPABASE_KEY = process.env.SUPABASE_SERVICE_ROLE_KEY || process.env.SUPABASE_KEY || process.env.STORAGE_KEY;

if (!SUPABASE_URL || !SUPABASE_KEY) {
    console.error("❌ Missing Supabase keys.");
    process.exit(1);
}

// 1. Setup Clients
const supabase = createClient(SUPABASE_URL, SUPABASE_KEY);
const prisma = new PrismaClient();

// Configure Imgly
const distPath = path.join(__dirname, 'dist') + path.sep;
const publicPath = 'file://' + distPath.replace(/\\/g, '/');
const imglyConfig = {
    publicPath: publicPath,
    debug: false,
    model_path: publicPath
};

async function main() {
    console.log("🚀 Starting Node.js Image Processing Pipeline (Prisma Edition)...");

    try {
        await prisma.$connect();
        console.log("✅ Connected to Database via Prisma.");

        // 2. Fetch Products
        // Query for valid products
        // main_image_url is required in schema, so we don't need to check for null
        const products = await prisma.product.findMany({
            select: {
                product_id: true,
                main_image_url: true
            }
        });

        console.log(`📦 Found ${products.length} products.`);

        // 3. Progress Bar
        const bar = new cliProgress.SingleBar({}, cliProgress.Presets.shades_classic);
        bar.start(products.length, 0);

        let successCount = 0;
        let failCount = 0;
        let skipCount = 0;

        for (const product of products) {
            const pid = product.product_id;
            const originalUrl = product.main_image_url;

            // Skip if already processed
            if (originalUrl.includes(SUPABASE_URL)) {
                skipCount++;
                bar.increment();
                continue;
            }

            try {
                // A. Download
                const response = await fetch(originalUrl);
                if (!response.ok) throw new Error(`Fetch failed: ${response.status} ${response.statusText}`);
                const blob = await response.blob();

                // B. Remove Background
                const processedBlob = await removeBackground(blob, imglyConfig);
                const processedBuffer = Buffer.from(await processedBlob.arrayBuffer());

                // C. Upload
                const fileName = `processed_${pid}_${Date.now()}.png`;
                const { data: uploadData, error: uploadError } = await supabase
                    .storage
                    .from('product')
                    .upload(fileName, processedBuffer, {
                        contentType: 'image/png',
                        upsert: true
                    });

                if (uploadError) throw new Error(`Upload error: ${uploadError.message}`);

                // Public URL
                const { data: publicUrlData } = supabase
                    .storage
                    .from('product')
                    .getPublicUrl(fileName);

                const newUrl = publicUrlData.publicUrl;

                // D. Update DB
                await prisma.product.update({
                    where: { product_id: pid },
                    data: { main_image_url: newUrl }
                });

                successCount++;

            } catch (err) {
                // console.error(`Failed ${pid}: ${err.message}`);
                failCount++;
            } finally {
                bar.increment();
            }
        }

        bar.stop();

        console.log("\n==========================================");
        console.log("✅ Processing Complete");
        console.log(`🟢 Successful: ${successCount}`);
        console.log(`🔴 Failed: ${failCount}`);
        console.log(`⚪ Skipped: ${skipCount}`);
        console.log("==========================================");

    } catch (err) {
        console.error("\n❌ Fatal Error:", err);
    } finally {
        await prisma.$disconnect();
    }
}

main();
